File: sponge-jruby/examples/script/rb/correlators.rb
# Sponge Knowledge base
# Using correlators

java_import org.openksavi.sponge.examples.SampleJavaCorrelator
java_import java.util.concurrent.atomic.AtomicInteger
java_import java.util.concurrent.atomic.AtomicBoolean

def onInit
    # Variables for assertions only
    $sponge.setVariable("hardwareFailureScriptCount", AtomicInteger.new(0))
    $sponge.setVariable("hardwareFailureJavaCount", AtomicInteger.new(0))
    $sponge.setVariable("hardwareFailureScriptFinishCount", AtomicInteger.new(0))
    $sponge.setVariable("hardwareFailureJavaFinishCount", AtomicInteger.new(0))
end

class SampleCorrelator < Correlator
    def onConfigure
        self.withEvents(["filesystemFailure", "diskFailure"]).withMaxInstances(1)
    end
    def onAcceptAsFirst(event)
        return event.name == "filesystemFailure"
    end
    def onInit
        @eventLog = []
    end
    def onEvent(event)
        @eventLog  << event
        self.logger.debug("{} - event: {}, log: {}", self.hashCode(), event.name, @eventLog.map(&:to_s))
        $sponge.getVariable("hardwareFailureScriptCount").incrementAndGet()
        if @eventLog.length >= 4
            $sponge.getVariable("hardwareFailureScriptFinishCount").incrementAndGet()
            self.finish()
        end
    end
end

def onLoad
    $sponge.enableJava(SampleJavaCorrelator)
end

def onStartup
    $sponge.event("filesystemFailure").set("source", "server1").send()
    $sponge.event("diskFailure").set("source", "server1").send()
    $sponge.event("diskFailure").set("source", "server2").send()
    $sponge.event("diskFailure").set("source", "server1").send()
    $sponge.event("diskFailure").set("source", "server2").send()
end

File: sponge-nashorn/examples/script/js/correlators.js
/**
 * Sponge Knowledge Base
 * Using correlators
 */

var AtomicInteger = java.util.concurrent.atomic.AtomicInteger;
var AtomicBoolean = java.util.concurrent.atomic.AtomicBoolean;

function onInit() {
    // Variables for assertions only
    sponge.setVariable("hardwareFailureScriptCount", new AtomicInteger(0));
    sponge.setVariable("hardwareFailureJavaCount", new AtomicInteger(0));
    sponge.setVariable("hardwareFailureScriptFinishCount", new AtomicInteger(0));
    sponge.setVariable("hardwareFailureJavaFinishCount", new AtomicInteger(0));
}

var SampleCorrelator = Java.extend(Correlator, {
    onConfigure: function(self) {
        self.withEvents(["filesystemFailure", "diskFailure"]).withMaxInstances(1);
    },
    onAcceptAsFirst: function(self, event) {
        return event.name == "filesystemFailure"
    },
    onInit: function(self) {
        self.target = new function() {
            this.eventLog = [];
        }
    },
    onEvent: function(self, event) {
        self.target.eventLog.push(event);
        self.logger.debug("{} - event: {}, log: {}", self.hashCode(), event.name, self.target.eventLog.toString());
        sponge.getVariable("hardwareFailureScriptCount").incrementAndGet();
        if (self.target.eventLog.length >= 4) {
            sponge.getVariable("hardwareFailureScriptFinishCount").incrementAndGet();
            self.finish();
        }
    }
});

function onLoad() {
    sponge.enableJava(org.openksavi.sponge.examples.SampleJavaCorrelator.class);
}

function onStartup() {
    sponge.event("filesystemFailure").set("source", "server1").send();
    sponge.event("diskFailure").set("source", "server1").send();
    sponge.event("diskFailure").set("source", "server2").send();
    sponge.event("diskFailure").set("source", "server1").send();
    sponge.event("diskFailure").set("source", "server2").send();
}